sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/HyperLogLogPlusPlusHelper.scala

```diff
@@ -93,6 +94,10 @@ class HyperLogLogPlusPlusHelper(relativeSD: Double) extends Serializable {
     val value = dataType match {
       case FloatType => FLOAT_NORMALIZER.apply(_value)
       case DoubleType => DOUBLE_NORMALIZER.apply(_value)
+      case s: StringType => _value match {
+        case v: UTF8String => CollationFactory.getCollationKeyBytes(
+          v, s.collationId)
+      }
```

Comment:
```suggestion
      case st: StringType => _value match {
        case val: UTF8String => CollationFactory.getCollationKeyBytes(val, st.collationId)
      }
```

Reply:
also, let's think about scala match errors - they can sometimes be daunting to debug later on...

Reply:
we can either attempt something like this to avoid `match` in outer `case`:
```
      case st: StringType if _value.isInstanceOf[UTF8String] =>
        CollationFactory.getCollationKeyBytes(_value, st.collationId)
```

or catch any potential mismatches manually like so:
```
      case st: StringType => _value match {
        case val: UTF8String => CollationFactory.getCollationKeyBytes(val, st.collationId)
        case _ => _value
      }
```

I'm leaning towards the first approach